### PR TITLE
Add search functionality and display page

### PR DESCRIPTION
**Pull request by nathanbegbie**

### CODE REVIEW

**Comment by gustavp** [comment]
This is a test by gustav to show jane how issues work.

**Comment by Rizziepit** [comment]
@miltontony I've updated this PR with the `get_object` / `to_object` changes

File: springboard/views.py [hunk 1]
```diff
@@ -74,3 +115,7 @@ def api_notify(self):
         fastforward.delay(self.workspace.working_dir,
                           self.workspace.index_prefix)
         return {}
+
+    def get_category(self, uuid):
+        [category] = self.workspace.S(Category).filter(uuid=uuid)
+        return category.get_object()
```
Comment (miltontony) [inline springboard/views.py]: 1 more `get_object` here

**Comment by Rizziepit** [comment]
@miltontony last `get_object` removed. thanks for that

**Comment by smn** [comment]
Does this still need a review?

**Comment by Rizziepit** [comment]
@smn not sure, but I'd like your input on this. This is the first stuff copied from unicore-cms and the structure might need to change to set a better precedent for springboard.

File: springboard/templates/molecules.html [hunk 2]
```diff
@@ -0,0 +1,123 @@
+{% import 'atoms.html' as atoms %}
+
+{% macro list_of_languages(view, languages, styling_classes) -%}
+<ul class="nav list_of_languages list-inline {{styling_classes}}" role="navigation">
+{% for lang_code, lang_name in languages %}
+  <li>
+    <a href={{view.request.route_url('locale_matched',  language=lang_code ) }}>{{lang_name}}</a>
```
Comment (smn) [inline springboard/templates/molecules.html]: There's a [`route_url`](https://github.com/universalcore/springboard/blob/develop/springboard/templates/home.jinja2#L24) template filter that may be nicer to use.

File: springboard/templates/molecules.html [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  </li>
+{% endfor %}
+	<li><a href="/spice/locale/change/">&gt;&gt;</a></li>
```
Comment (smn) [inline springboard/templates/molecules.html]: What is `spice`? It's not matched on any routes. Copy paste error from somewhere?

File: springboard/templates/molecules.html [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    <p>{{num_of_results}} search results for: <b>{{query}}</b></p>
+  {% endif %}
+{%- endmacro %}
+
+{% macro article_summary(result, view) -%}
+  {% if result.primary_category  %}
+      <p>{{view.get_category(result.primary_category).title}}</p>
+      {% endif %}
+      <a href={{view.request.route_url('page', uuid=result.uuid)}}>{{result.title}} ... Read More &gt;</a>
+{%- endmacro %}
```
Comment (smn) [inline springboard/templates/molecules.html]: missing new line.

Comment (smn) [inline springboard/templates/molecules.html]: There's a bunch of english placeholder text in this that we probably need to wrap in `{% trans %}` kind of stuff.

File: springboard/templates/search_results.jinja2 [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+	{% endif %}
+
+	{{organisms.list_of_returned_articles(paginator, view)}}
+
+{% else %}
+	{{molecules.message_for_no_results(query)}}
+{% endif %}
+
+</div>
+{% endblock %}
```
Comment (smn) [inline springboard/templates/search_results.jinja2]: missing new line.

File: springboard/templates/molecules.html (hunk 3 above)
Comment (miltontony) [inline springboard/templates/molecules.html]: Copy paste error from `unicore-cms`. This should be `{{view.request.route_url('locale_change')}}` (or something)

File: springboard/tests/test_search.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        resp = self.app.get('/search/', params={'q': 'sample'}, status=200)
+        self.assertTrue(
+            'Next&nbsp;&gt;</a>'
+            in resp.body)
+
+    def test_search_profanity(self):
+        self.app = self.mk_app(self.workspace)
+        self.mk_pages(self.workspace, count=2)
+
+        resp = self.app.get('/search/', params={'q': 'kak'}, status=200)
```
Comment (smn) [inline springboard/tests/test_search.py:65]: What are we testing here exactly?

File: springboard/templates/molecules.html (hunk 3 above)
Comment (smn) [inline springboard/templates/molecules.html]: `{{'locale_change'|route_url}}`

File: springboard/utils.py [hunk 7]
```diff
@@ -38,3 +39,114 @@ def wrapper(self, *args, **kwargs):
...
+        return self.es_results.__len__()
+
+    def __getitem__(self, k):
+        if isinstance(k, slice):
+            return ResultGenerator(self.es_results.__getitem__(k))
+        return self.es_results.__getitem__(k).to_object()
+
+
+def to_eg_objects(es_results):
+    return ResultGenerator(es_results)
```
Comment (smn) [inline springboard/utils.py]: This, `to_eg_objects` and `ResultGenerator` seem like they're more complex than they need to be. Can we not do this simpler in the `EGPaginator.get_page()` function?

**Comment by smn** [comment]
Sorry, very late review on some initial work that I think @nathanbegbie did ages ago.
@nathanbegbie any chance you're able to work on some of the stuff raised?

File: springboard/utils.py (hunk 7 above)
Comment (Rizziepit) [inline springboard/utils.py]: +1 on this. `to_eg_objects` is used for other stuff in unicore-cms but we might never get around to needing it in springboard.

**Comment by nathanbegbie** [comment]
@smn @Rizziepit, sorry, was working on this before my 2 week hiatus. Ran into issues with locales and getting certain search tests related to language to work, so there's now this; https://github.com/universalcore/springboard/tree/feature/issue-21-add-locale-change which I wanted to finish, merge into develop and then pull into this branch. Will get back on this after lunch.

**Comment by smn** [comment]
@nathanbegbie cool, I turned #21 into a pull request so we can review that one first.

**Comment by Rizziepit** [comment]
@nathanbegbie can you open a PR just for the pagination parts? I need pagination for https://github.com/universalcore/springboard-iogt, plus I think pagination will benefit from being reviewed separately.

**Comment by nathanbegbie** [comment]
@Rizziepit I'm exam-ing at the mo, I can get back on this on Thursday ~12pm. Is that too late?

**Comment by Rizziepit** [comment]
@nathanbegbie I forgot about exams. No prob. I'm going to go ahead and split this up then, and land the pagination part.

**Comment by Rizziepit** [comment]
@nathanbegbie pagination has been landed and I merged it back into this branch. `EGPaginator` has been replaced with `Paginator` and that's about it.

**Comment by nathanbegbie** [comment]
@Rizziepit @miltontony @smn Ready for review

**Comment by smn** [comment]
:+1:

**Pull request merged**